### ISSUE TITLE
Refactoring: Use namespace-based APIs instead of Tenant-based APIs

### ISSUE THREAD

**Issue by erlemantos** [eclipse/vorto#2152]
- A lot of UI code rely on the tenant-based APIs which we are deprecating to make way for the namespace-based APIs

**Comment by kolotu** [comment]
Hi @erlemantos did you already start with this issue, or is this something that we should pick up?

**Comment by erlemantos** [comment]
@kolotu I haven't started this issue yet.

**Comment by mena-bosch** [comment]
Hi @erlemantos , I am having a little trouble narrowing down what to do here. 
Could you specify what exactly you deem necessary?
Is it just cosmetics (renaming API methods, variables, presentational values) or is there more to it?
What are the areas impacted / caveats you can think of?

Thanks in advance.

**Comment by erlemantos** [comment]
Hi @mena-bosch ,

Currently, we have a TenantManagementController that contains handlers for REST calls of the form "/rest/tenants/<tenantId>/...". This tenantId is generated and not user-friendly. We were also trying to move away from the "tenant" terminology and into "namespace". The plan was to remove all the services that rely on tenantId (/rest/tenants/<tenantId>/...) and port all those services to namespace (rest/namespaces/<namespace>/...). 

In order to do this without generating much impact, we were planning to do it on the UI first by removing all frontend rest calls to /rest/tenants/<tenantId>/...) and replace those calls with their /rest/namespaces/<namespace>/... counterpart.

When we have removed all the frontend rest calls to "/rest/tenants/<tenantId>/...", then maybe we can already remove the tenant-based services in TenantManagementController.

This task is about the first step, which is to remove all frontend calls to "/rest/tenants/<tenantId>/..." and convert them to the namespace-based API's (/rest/namespaces/<namespace>/...).

Now, if a tenant-based service doesn't exist yet in the namespace form, then an equivalent service in namespace form should be created.

**Comment by mena-bosch** [comment]
Thanks for the clarification @erlemantos . The way I see it, it is *nearly* "cosmetics" insofar as we'd just be changing names but keeping idempotent functionality then, with the exception of the tenant ID that could be hopefully just replaced by the namespace's name. 
Either way, I think this explanation gives me enough meat to get started on this one, so I'll be getting my hands dirty soon. 
May still get back to you with more questions once I have something concrete if you don't mind.

**Comment by mena-bosch** [comment]
Had to take a break from this one due to more pressing issues. 

Today @kolotu and I had a quick brainstorm on how to start with this. 

The initial idea would be to:

1. Create a new controller handling all new "namespaces" endpoints, converted from the functionality currently implemented in the 
   - AccountController
   - TenantManagementController
   - BackupController

2. The new controller would autowire the tenant controllers above and act as a proxy
3. Then the UI and back-end tests/filters could be changed to use the new controller

Once this is done, phase 2 would imply actions such as:

- Re-designing the relevant DB table(s)/model and the tenant-specific functionality
- Removing the old tenant functionality and controllers altogether
- Documenting in the migration markdown

As this looks like a lot of changes, it might be desirable to implement phase 2 in a different task.

**Comment by mena-bosch** [comment]
### Security-ish finding

The current tenantID generation is **in the UI**. 
It's a (debatable) piece of Javascript code that generates a semi-random ASCII string deemed long enough to avoid collisions with existing ones (although there is no real pathway if a collision occurs - I guess the end user would see this as a glitch and start over). 

While Spring security seems to do a good job preventing requests with SQL injection characters from going through, this methodology still leaves knowledgeable users with the option to tamper with the generated tenant ID and use their own, e.g. instead of `PUT /rest/tenants/[some random ASCII]` you can tamper the request and end up with `PUT /rest/tenants/iliketurtles`.

That will need to change sooner than later.

**Comment by mena-bosch** [comment]
**Update**

I have been working on the angular vs java controller communication and have two new endpoints in the NamespaceController:

- PUT with empty body + namespace as path variable to try and create a new namespace
- DELETE + namespace as path variable to try and delete an existing namespace

The validation and permission checks in the back-end are adapted to follow the same rules as before, with some embellishments when it comes to responses vs just throwing exception at times. 

Under the hood, the controller populates all the (redundant?) information the tenant service needs from context - some of which is directly passed as `null` (e.g. the tenant's authentication provider). 

The real challenge now is the GET endpoints e.g. the one to retrieve all namespaces: the array of "tenants" is used all over the place in the angular code, so it won't be trivial to convert it to a collection of namespace objects instead, given the cross-dependencies, weak types, and lack of proper tooling to manage semi-automatic refactory the way one would do with Java. 

I've started by renaming the tenantManagement controller, HTML fragment and some variables. 

Will continue in that direction next week, although this feels a lot like a ticking regression bomb.

**Comment by mena-bosch** [comment]
Added one property to the NamespaceDTO to feature the `admins` collection - the rest of the properties of the original Tenant can likely be removed safely in the front-end. 

One remark: the existing, new NamespaceController only featured two endpoints when I started working on it - both to retrieve namespaces and both taking a `Principal` as parameter. 

Not sure if there's some Spring magic to inject that but it can easily be inferred from the session and moved out of the method's signature. 

I think it could be kept there _only_ if there is any future use case to retrieve namespaces by a specific user (instead of the user logged on). 

Will "overload" the `getNamespaces` one for now to continue working safely (and not break the existing tests) although for purposes of cleanliness, I'd rather get rid of the `Principal` parametrization at some point if possible. 

@erlemantos any thoughts?
